File: app/src/main/java/com/example/expense_management/Database/Dao.java
package com.example.expense_management.Database;

import androidx.lifecycle.LiveData;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@androidx.room.Dao
public interface Dao {

    @Insert(onConflict =OnConflictStrategy.REPLACE)
    void insert(DataItems items);

    @Delete
    void delete(DataItems items);

    @Update
    void update(DataItems item);

    @Query("UPDATE EXPENSES SET DATE =:uDate WHERE date =:date")
    void updateDate(long uDate,long date);

    @Query("SELECT date FROM EXPENSES ORDER BY DATE")
    List<Long> getAllDate();

    @Query("SELECT COUNT(date) FROM EXPENSES")
    Integer getCount();

    @Query("SELECT * FROM EXPENSES ORDER BY DATE")
    LiveData<List<DataItems>> getAllAscending();

    @Query("SELECT *FROM EXPENSES ORDER BY DATE DESC")
    LiveData<List<DataItems>> getAllDescending();

    @Query("SELECT *FROM EXPENSES ORDER BY grossMoneyGot DESC")
    LiveData<List<DataItems>> getAllGotDescending();

    @Query("SELECT *FROM EXPENSES ORDER BY grossMoneyExpense DESC")
    LiveData<List<DataItems>> getAllExpenseDescending();

    @Query("SELECT * FROM EXPENSES ORDER BY grossMoneyGot")
    LiveData<List<DataItems>> getAllGotAscending();

    @Query("SELECT * FROM EXPENSES ORDER BY grossMoneyExpense")
    LiveData<List<DataItems>> getAllExpenseAscending();
    @Query("select *from expenses where date=:date ")
    LiveData<DataItems> getRow(long date);

    @Query("select *from expenses where date=:date ")
    DataItems getRoww(long date);
}

File: app/src/main/java/com/example/expense_management/database/Dao.kt
package com.example.expense_management.database

import androidx.lifecycle.LiveData
import androidx.room.*
import androidx.room.Dao

@Dao
interface Dao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(items: DataItems?)

    @Delete
    suspend fun delete(items: DataItems?)

    @Update
    suspend fun update(item: DataItems?)

    @Query("UPDATE EXPENSES SET DATE =:uDate WHERE date =:date")
    suspend fun updateDate(uDate: Long, date: Long)

    @Query("SELECT date FROM EXPENSES ORDER BY DATE")
    suspend fun  allDate(): List<Long>

    @Query("SELECT COUNT(date) FROM EXPENSES")
    fun  count(): Int?

    @Query("SELECT * FROM EXPENSES ORDER BY grossMoneyExpense")
    fun dataExpenseAscending() : LiveData<List<DataItems>>

    @Query("SELECT * FROM EXPENSES ORDER BY grossMoneyGot")
    fun dataGotAscending() : LiveData<List<DataItems>>

    @Query("SELECT * FROM EXPENSES ORDER BY grossMoneyExpense DESC")
    fun dataExpenseDescending() : LiveData<List<DataItems>>

    @Query("SELECT * FROM EXPENSES ORDER BY grossMoneyGot DESC")
    fun dataGotDescending() : LiveData<List<DataItems>>

    @Query("SELECT * FROM EXPENSES ORDER BY DATE DESC")
    fun dataDescending() : LiveData<List<DataItems>>


    @Query("SELECT * FROM EXPENSES ORDER BY DATE")
    fun dataAscending() : LiveData<List<DataItems>>

    @Query("select *from expenses where date=:date ")
    fun getRow(date: Long): LiveData<DataItems?>?

    @Query("select *from expenses where date=:date ")
    suspend fun getRoww(date: Long?): DataItems?

    @Query("SELECT Sum(grossMoneyExpense) FROM EXPENSES WHERE date BETWEEN :first And :last ")
    suspend fun getExpense(first: Long, last: Long): Int

    @Query("SELECT Sum(grossMoneyGot) FROM EXPENSES WHERE date BETWEEN :first And :last ")
    suspend fun getGain(first: Long, last: Long): Int
}